src/index.js

```diff
@@ -0,0 +1,143 @@
+'use strict'
+
+const Emittery = require('emittery')
+const debug = require('debug')
+
+const PeerId = require('peer-id')
+const PeerInfo = require('peer-info')
+const randomBytes = require('libp2p-crypto/src/random-bytes')
+const multiaddr = require('multiaddr')
+
+const PB = require('./query')
+
+const log = debug('libp2p:discovery:pubsub')
+log.error = debug('libp2p:discovery:pubsub:error')
+const TOPIC = '_peer-discovery._p2p._pubsub'
+
+/**
+ * @typedef Message
+ * @property {string} from
+ * @property {Buffer} data
+ * @property {Buffer} seqno
+ * @property {Array<string>} topicIDs
+ * @property {Buffer} signature
+ * @property {key} Buffer
+ */
+
+/**
+  * A Peer Discovery Service that leverages libp2p Pubsub to find peers.
+  */
+class PubsubPeerDiscovery extends Emittery {
+  // TODO: The default delay may be too high here.
+  // By the time the service is started, libp2p will be listening on all transports (including relays)
+  // and pubsub will be started. The delay is to give pubsub time to prime. Ideally, we might do an
+  // initial check to wait for other subscribers before making our startup Query. However, in the
+  // unlikely event that we are the first peer, we would be waiting unncessarily, as the next peer would
+  // query and result in our discovery of them
+  /**
+   *
+   * @param {Libp2p} param0.libp2p Our libp2p node
+   * @param {Number} param0.delay How long to wait (ms) after startup before publishing our Query
+   */
+  constructor ({ libp2p, delay = 5000 }) {
+    super()
+    this.libp2p = libp2p
+    this.delay = delay
+    this._timeout = null
+  }
+
+  /**
+   * Subscribes to the discovery topic on `libp2p.pubsub` and peforms a query
+   * after `this.delay` milliseconds
+   */
+  start () {
```

Comment:
validate that this was not already started by checking the `this._timeout` value